File: shared/src/types/campaign.ts
import type { MarketplaceChannel } from './enums';

/**
 * Campanhas de divulgação social (Facebook/Instagram): orgânicas (posts
 * agendados em lote, sem custo) ou pagas (Facebook Marketing API, com gasto
 * real). Um único tipo com discriminador `type`, no mesmo espírito do
 * `Product` — campos fixos + bloco solto por variante.
 */
export const CampaignStatus = {
  DRAFT: 'draft',
  ACTIVE: 'active',
  PAUSED: 'paused',
  COMPLETED: 'completed',
  ARCHIVED: 'archived',
} as const;
export type CampaignStatus = (typeof CampaignStatus)[keyof typeof CampaignStatus];

export interface OrganicCampaignItem {
  productId: string;
  scheduledFor: string; // ISO date
  status: 'queued' | 'sent_for_approval' | 'posted' | 'skipped';
}

export interface OrganicCampaignConfig {
  channels: MarketplaceChannel[];
  intervalDays: number;
  startDate: string; // ISO date
  /** Materializado na criação da campanha — lista fixa, não recalculada depois. */
  items: OrganicCampaignItem[];
}

export interface PaidCampaignTargeting {
  countries: string[];
  ageMin: number;
  ageMax: number;
  genders?: ('male' | 'female')[];
}

export interface PaidCampaignConfig {
  objective: 'POST_ENGAGEMENT' | 'REACH' | 'TRAFFIC';
  dailyBudgetCents: number;
  currency: string;
  targeting: PaidCampaignTargeting;
  productId: string;
  channel: MarketplaceChannel;
  external: {
    campaignId?: string;
    adSetId?: string;
    adId?: string;
    creativeId?: string;
  };
  startDate?: string;
  endDate?: string;
  /** Preenchido se a cadeia de criação na Marketing API falhou no meio do caminho. */
  lastError?: string;
}

export interface Campaign {
  id: string;
  ownerId: string;
  type: 'organic' | 'paid';
  name: string;
  status: CampaignStatus;
  organic?: OrganicCampaignConfig;
  paid?: PaidCampaignConfig;
  createdAt: string;
  updatedAt: string;
}

File: backend/app/shared/schemas/campaign.py
from enum import StrEnum

from pydantic import BaseModel

from app.shared.schemas.enums import MarketplaceChannel


class CampaignStatus(StrEnum):
    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class OrganicCampaignItemStatus(StrEnum):
    QUEUED = "queued"
    SENT_FOR_APPROVAL = "sent_for_approval"
    POSTED = "posted"
    SKIPPED = "skipped"


class OrganicCampaignItem(BaseModel):
    productId: str
    scheduledFor: str
    status: OrganicCampaignItemStatus


class OrganicCampaignConfig(BaseModel):
    channels: list[MarketplaceChannel]
    intervalDays: int
    startDate: str
    items: list[OrganicCampaignItem]


class PaidCampaignGender(StrEnum):
    MALE = "male"
    FEMALE = "female"


class PaidCampaignObjective(StrEnum):
    POST_ENGAGEMENT = "POST_ENGAGEMENT"
    REACH = "REACH"
    TRAFFIC = "TRAFFIC"


class PaidCampaignTargeting(BaseModel):
    countries: list[str]
    ageMin: int
    ageMax: int
    genders: list[PaidCampaignGender] | None = None


class PaidCampaignExternalIds(BaseModel):
    campaignId: str | None = None
    adSetId: str | None = None
    adId: str | None = None
    creativeId: str | None = None


class PaidCampaignConfig(BaseModel):
    objective: PaidCampaignObjective
    dailyBudgetCents: int
    currency: str
    targeting: PaidCampaignTargeting
    productId: str
    channel: MarketplaceChannel
    external: PaidCampaignExternalIds
    startDate: str | None = None
    endDate: str | None = None
    lastError: str | None = None


class CampaignType(StrEnum):
    ORGANIC = "organic"
    PAID = "paid"


class Campaign(BaseModel):
    id: str
    ownerId: str
    type: CampaignType
    name: str
    status: CampaignStatus
    organic: OrganicCampaignConfig | None = None
    paid: PaidCampaignConfig | None = None
    createdAt: str
    updatedAt: str
